chain-impl-mockchain/src/transaction/transaction.rs

```diff
@@ -261,55 +264,57 @@ pub(super) struct TransactionStruct {
 /// Verify the structure of the transaction and return all the offsets
 fn get_spine<P: Payload>(slice: &[u8]) -> Result<TransactionStruct, TransactionStructError> {
     let sz = slice.len();
-    let mut rb = ReadBuf::from(slice);
+    let mut codec = Codec::new(slice);
 
     // read payload
     if P::HAS_DATA {
-        P::read_validate(&mut rb).map_err(|_| TransactionStructError::PayloadInvalid)?;
+        P::deserialize_validate_from_slice(&mut codec)
+            .map_err(|_| TransactionStructError::PayloadInvalid)?;
     }
 
     // read date
-    let epoch = rb
-        .get_u32()
+    let epoch = codec
+        .get_be_u32()
         .map_err(|_| TransactionStructError::CannotReadExpiryDate)?;
-    let slot_id = rb
-        .get_u32()
+    let slot_id = codec
+        .get_be_u32()
         .map_err(|_| TransactionStructError::CannotReadExpiryDate)?;
     let valid_until = BlockDate { epoch, slot_id };
 
     // read input and outputs
-    let nb_inputs = rb
+    let nb_inputs = codec
         .get_u8()
         .map_err(|_| TransactionStructError::CannotReadNbInputs)?;
-    let nb_outputs = rb
+    let nb_outputs = codec
         .get_u8()
         .map_err(|_| TransactionStructError::CannotReadNbOutputs)?;
 
-    let inputs_pos = rb.position();
-    rb.skip_bytes(nb_inputs as usize * INPUT_SIZE)
-        .map_err(|_| TransactionStructError::InputsInvalid)?;
-    let outputs_pos = rb.position();
+    let inputs_pos = sz - codec.bytes_left();
+    codec.skip_bytes(nb_inputs as usize * INPUT_SIZE);
+    let outputs_pos = sz - codec.bytes_left();
     for _ in 0..nb_outputs {
-        Output::<Address>::read_validate(&mut rb)
+        Output::<Address>::deserialize_validate(&mut codec)
             .map_err(|_| TransactionStructError::OutputsInvalid)?;
     }
 
     // read witnesses
-    let witnesses_pos = rb.position();
+    let witnesses_pos = sz - codec.bytes_left();
     for _ in 0..nb_inputs {
-        Witness::read_validate(&mut rb).map_err(|_| TransactionStructError::WitnessesInvalid)?;
+        Witness::deserialize_validate_from_slice(&mut codec)
+            .map_err(|_| TransactionStructError::WitnessesInvalid)?;
     }
 
     // read payload auth
-    let payload_auth_pos = rb.position();
+    let payload_auth_pos = sz - codec.bytes_left();
     if P::HAS_DATA && P::HAS_AUTH {
-        if rb.is_end() {
+        if codec.bytes_left() == 0 {
```

Comment:
This value is routinely compared to zero. It looks like a good candidate to be wrapped into a function like `fn has_bytes_left(&self) -> bool`.